### PR TITLE
[LTC] Refactor torch_lazy_tensors::Device

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request will keep track of the progress of refactoring torch_lazy_tensors::Device.
1. The first commit removed torch_lazy_tensors::DeviceType from all places and added a HardwareType type as an extension point for backends that need this type of information.
2. Clean up class Device interface.
3. Remove most use cases of string devices.
4. Introduce BackendDeviceType to class Device.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

To be noted:
It's easier to review it commit by commit.

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/device.h
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/device.h:12]: @JackCaoG Here is the new design. This aiming of this new BackendDeviceType is to provide an extensible way for backends to pretty print device type while keeping the many-to-one design philosophy of class Device intact and avoid changing too many code.

Why we need to pretty print Device? It's because we have (and want to offer) APIs that allow users to query the current hardware.

**Review by alanwaketan — COMMENTED** [review]
The pull request is ready for review.

**Review by wconstab — APPROVED** [review]
Thanks @alanwaketan looks great.  Let's wait for @JackCaoG to review.

For naming, Do you think it makes sense to use `BackendDevice`, `BackendData` (my other PR), `BackendInterface`?  I wanted a way to make it very clear these generic sounding objects are from the backend rather than the similar ones in pytorch..  Other option might be putting them in a `backend:: namespace`.

**Comment by alanwaketan** [comment]
> Thanks @alanwaketan looks great. Let's wait for @JackCaoG to review.
> 
> For naming, Do you think it makes sense to use `BackendDevice`, `BackendData` (my other PR), `BackendInterface`? I wanted a way to make it very clear these generic sounding objects are from the backend rather than the similar ones in pytorch.. Other option might be putting them in a `backend:: namespace`.

I think Backend* types make sense, and should probably stay in torch::lazy. The reason why I didn't make the name change is that I recognize that I need to do the renaming again after merging master. So, I would just leave the renaming work then.

**Review by JackCaoG — COMMENTED** [review]
Mostly LGTM

File: lazy_tensor_core/lazy_tensor_core/csrc/device.h
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/device.h:30]: FB style...

File: lazy_tensor_core/lazy_tensor_core/csrc/ir_dump_util.h
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ir_dump_util.h:9]: Using forward declarations instead of include in headers can help reduce build time.

File: lazy_tensor_core/lazy_tensor_core/csrc/lazy_graph_executor.cpp
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/lazy_graph_executor.cpp:439]: We currently haven't finalized the backend API to do string -> device mapping yet. Let me add a TODO and change the signatures to take `c10::ArrayRef<Device>`.

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/lazy_graph_executor.cpp:439]: Yea, you can just leave a TODO for now and I can add this back in with my PR that changes the backend interfaces.  Or whatever is easiest for you.

**Review by JackCaoG — APPROVED** [review]
Thanks!

**Comment by alanwaketan** [comment]
@wconstab @JackCaoG Thanks for approving the pull request. I will merge it once the CI passes.

**Pull request merged**
